app/models/tag.rb

```diff
@@ -674,302 +654,99 @@ def recategorize(new_type)
-      if old_merger && old_merger.canonical?
-        self.async(:remove_filter_taggings, old_merger.id)
-      end
-    end
-  end
-
-  # Add filter taggings for a given tag
-  # This is currently called only if this tag has just become canonical
-  def add_filter_taggings
-    # the "filter" method gets either this tag itself or its merger -- in practice will always be this tag because
-    # this method only gets called when this tag is canonical and therefore cannot have a merger
-    filter_tag = self.filter
-    return unless filter_tag && !filter_tag.new_record?
-
-    # we collect tags for resetting count so that it's only done once after we've added all filters to works
-    tags_that_need_filter_count_reset = []
-    items = self.works + self.external_works
-    items.each do |item|
-      if item.filters.include?(filter_tag)
-        # If the item filters already included the filter tag (e.g. because the
-        # new filter tag is a meta tag of an existing tag) we make sure to set
-        # the inheritance to false, since the item is now directly tagged with
-        # the filter or one of its synonyms
-        ft = item.filter_taggings.where(["filter_id = ?", filter_tag.id]).first
-        ft.update_attribute(:inherited, false)
-      else
-        FilterTagging.create(
-          filter: filter_tag,
-          filterable: item
-        )
-        # As of Rails 5 upgrade, this triggers a stack level too deep error
-        # because it triggers the `before_update
-        # :update_filters_for_canonical_change` callback. In 4.2 and before,
-        # after this point `canonical_changed?` has been reset and returns
-        # false. Now it returns true and causes an endless loop.
-        #
-        # Reference: https://github.com/rails/rails/issues/28908
-        #
-        # TODO: Keep an eye on this issue. We should not have to create the
-        # FilterTagging directly.
-        #
-        # work.filters << filter_tag
-        unless item.is_a?(ExternalWork) || tags_that_need_filter_count_reset.include?(filter_tag)
-          tags_that_need_filter_count_reset << filter_tag
-        end
-      end
-      unless filter_tag.meta_tags.empty?
-        filter_tag.meta_tags.each do |m|
-          next if item.filters.include?(m)
-          item.filter_taggings.create!(inherited: true, filter_id: m.id)
-          unless item.is_a?(ExternalWork) || tags_that_need_filter_count_reset.include?(m)
-            tags_that_need_filter_count_reset << m
-          end
-        end
+  # When canonical or merger_id changes, only the items directly tagged with
+  # this tag need their filters updated, so we queue up a call to
+  # update_filters_for_taggables after commit.
+  #
+  # Note that when a tag becomes non-canonical, all of its filter-taggings need
+  # to be deleted. But when a tag becomes non-canonical, all of its mergers and
+  # sub-tags will be deleted, which will result in the necessary items having
```

Comment:
Yeah. Calling `update_filters_for_filterables` on the main tag and `update_filters_for_taggables` on all of its synonyms would cause all of the works tagged with one of the synonyms to be processed twice.